File: 29.divide-two-integers.py
class Solution:
    def divide(self, dividend, divisor):
        """
        :type dividend: int
        :type divisor: int
        :rtype: int
        """
        sign = (dividend < 0) != (divisor < 0)
        ret = 0
        dividend = abs(dividend)
        divisor = abs(divisor)
        while dividend >= divisor:
            n = 1
            div = divisor
            while dividend >= div + div:
                div = div + div
                n = n + n
            dividend -= div
            ret += n
        if sign:
            ret = -ret
        if ret < -2**31 or ret > 2**31-1:
            return 2**31-1
        else:
            return ret

File: 29.divide-two-integers.js
/**
 * @param {number} dividend
 * @param {number} divisor
 * @return {number}
 */
var divide = function(dividend, divisor) {
    let sign = (dividend < 0) ^ (divisor < 0);
    let ret = 0;
    dividend = Math.abs(dividend);
    divisor = Math.abs(divisor);
    while (dividend >= divisor) {
        let n = 1;
        let div = divisor;
        while (dividend >= div + div) {
            div = div + div;
            n = n + n;
        }
        dividend -= div;
        ret += n;
    }
    if (sign) {
        ret = -ret;
    }
    if (ret < -Math.pow(2, 31) || ret > Math.pow(2, 31) - 1) {
        return Math.pow(2, 31) - 1;
    } else {
        return ret;
    }
};